### PR TITLE
Authentication refactoring: generalize account settings

### PR DESCRIPTION
**Pull request by rhymes**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

In this PR we generalize the last (hopefully) remaining part of the user facing code, the settings/account panel.

- renamed `remove_association` to `remove_identity`
- made sure one can only remove enabled providers (we don't show disabled providers anyway, but it's good to check server side as well)
- added `settings_url` which points to the URL where the user can land to remove the oauth association
- future proof some code and tests which checked if the total number of identities was 2 (we'll soon have more)
- use `sign_in_path` everywhere

## Related Tickets & Documents

Closes https://github.com/thepracticaldev/dev.to/issues/7585

## Added tests?

- [x] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -130,22 +132,34 @@ def full_delete
...
 
-    if @user.identities.size == 2 && identity
+    error_message = "An error occurred. Please try again or send an email to: #{SiteConfig.email_addresses[:default]}"
+    unless Authentication::Providers.enabled?(params[:provider])
+      flash[:error] = error_message
+      redirect_to user_settings_path(@tab)
+      return
+    end
+
+    provider = Authentication::Providers.get!(params[:provider])
```
Comment (mstruve) [inline app/controllers/users_controller.rb:145]: could use your new helper here `authentication_provider(provider_name)`

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -130,22 +132,34 @@ def full_delete
...
+
+    identity = @user.identities.find_by(provider: provider.provider_name)
+
+    if identity && @user.identities.size > 1
       identity.destroy
 
-      identity_username = "#{provider}_username".to_sym
-      @user.update(identity_username => nil, :profile_updated_at => Time.current)
+      @user.update(
+        provider.user_username_field => nil,
```
Comment (mstruve) [inline app/controllers/users_controller.rb:153]: Should this be `provider.class.user_username_field` like it is [below in the authenticator class](https://github.com/thepracticaldev/dev.to/pull/7597/files#diff-863cabff965798c30566df1bb5e2b3a8R99)?

File: app/views/users/_account.html.erb [hunk 3]
```diff
@@ -45,7 +45,7 @@
 <div class="crayons-box crayons-box--danger p-6 grid gap-6">
   <h2 class="color-accent-danger">Danger Zone</h2>
 
-  <% if @user.identities.size == 2 %>
+  <% if @user.identities.size > 1 %>
```
Comment (mstruve) [inline app/views/users/_account.html.erb:48]: 🔥

File: app/controllers/users_controller.rb (hunk 1 above)
Comment (citizen428) [reply app/controllers/users_controller.rb:145]: It's a view helper, so you'd need to call it as`helpers.authentication_provider(params[:provider])`.

File: app/services/authentication/providers/github.rb [hunk 4]
```diff
@@ -25,10 +26,22 @@ def existing_user_data
         }
       end
 
+      def self.user_created_at_field
```
Comment (citizen428) [inline app/services/authentication/providers/github.rb:29]: Tbh I liked them better when they were instance methods, as you always seem to be calling them from an instance context as illustrated by the `provider.class.user_created_at_field` calls. Ruby is not quite as strict as Python about making things static when you don't access `self` in the method body. As a sort of compromise you can leave them as is and add 

```ruby
delegate :user_created_at_field, to: :class
```

Which will keep them on the class but make them a bit less awkward to acess from an instance.

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (rhymes) [reply app/controllers/users_controller.rb:153]: it's a bit scary that it didn't fail any test 👀 
Going to see why

File: app/services/authentication/providers/github.rb (hunk 4 above)
Comment (rhymes) [reply app/services/authentication/providers/github.rb:29]: Awesome!

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (rhymes) [reply app/controllers/users_controller.rb:153]: Oh I know why, that's the class object, not the instance object, that's why it works :D I'm going to apply @citizen428 suggestion here https://github.com/thepracticaldev/dev.to/pull/7597#discussion_r417809095 so we don't get confused ;-)

File: app/services/authentication/providers/provider.rb [hunk 5]
```diff
@@ -2,12 +2,8 @@ module Authentication
   module Providers
     # Authentication provider
     class Provider
-      # name of the field to store the upstream identity's creation date in
-      CREATED_AT_FIELD = "".freeze
-      # name of the field to store the upstream identity's username in
-      USERNAME_FIELD = "".freeze
-
       delegate :email, :nickname, to: :info, prefix: :user
+      delegate :user_created_at_field, :user_username_field, to: :class
```
Comment (mstruve) [inline app/services/authentication/providers/provider.rb:6]: 🔥

**Pull request merged**